Change server to runtime or instance (#316)
OpenLiberty/draft-guide-gettingstarted: README.adoc

You'll also explore how to package your application with Open Liberty so that it can be deployed anywhere in one go. You will then make Liberty configuration and code changes and see how they are immediately picked up by a running instance.

> Finally, you will package the application along with the runtime configuration into a Docker image and run that image as a container.

Reviewer: ```suggestion
Finally, you will package the application along with Liberty's configuration into a Docker image and run that image as a container.
```

Your application is configured to be built with Maven. Every Maven-configured project contains a [hotspot]`pom.xml` file, which defines the project configuration, dependencies, plug-ins, and so on.

> Your [hotspot]`pom.xml` file is located in the `start` directory and is configured to include the [hotspot=libertyMavenPlugin]`liberty-maven-plugin`, which allows you to install applications into Open Liberty and manage the runtime instances.

Reviewer: ```suggestion
Your [hotspot]`pom.xml` file is located in the `start` directory and is configured to include the [hotspot=libertyMavenPlugin]`liberty-maven-plugin`, which allows you to install applications into Open Liberty and manage the associated Liberty instances.
```

The `mvn` command initiates a Maven build, during which the `target` directory is created to store all build-related files.

> The `liberty:run` argument specifies the Open Liberty `run` goal, which starts an Open Liberty instance in the foreground. As part of this phase, an Open Liberty runtime is downloaded and installed into the `target/liberty/wlp` directory, a runtume instance is created and configured in the `target/liberty/wlp/usr/servers/defaultServer` directory, and the application is installed into that instance via https://www.ibm.com/support/knowledgecenter/en/SSEQTP_liberty/com.ibm.websphere.wlp.doc/ae/rwlp_loose_applications.html[loose config^].

Reviewer: ```suggestion
The `liberty:run` argument specifies the Open Liberty `run` goal, which starts an Open Liberty instance in the foreground. As part of this phase, an Open Liberty runtime is downloaded and installed into the `target/liberty/wlp` directory, a Liberty instance is created and configured in the `target/liberty/wlp/usr/servers/defaultServer` directory, and the application is installed into that instance via https://www.ibm.com/support/knowledgecenter/en/SSEQTP_liberty/com.ibm.websphere.wlp.doc/ae/rwlp_loose_applications.html[loose config^].
```

> The `liberty:run` argument specifies the Open Liberty `run` goal, which starts an Open Liberty instance in the foreground. As part of this phase, an Open Liberty runtime is downloaded and installed into the `target/liberty/wlp` directory, a Liberty instance is created and configured in the `target/liberty/wlp/usr/servers/defaultServer` directory, and the application is installed into that instance via https://www.ibm.com/support/knowledgecenter/en/SSEQTP_liberty/com.ibm.websphere.wlp.doc/ae/rwlp_loose_applications.html[loose config^].

Reviewer 2: ```suggestion
The `liberty:run` argument specifies the Open Liberty `run` goal, which starts an Open Liberty instance in the foreground. As part of this phase, an Open Liberty runtime is downloaded and installed into the `target/liberty/wlp` directory, an instance of Liberty is created and configured in the `target/liberty/wlp/usr/servers/defaultServer` directory, and the application is installed into that instance via https://www.ibm.com/support/knowledgecenter/en/SSEQTP_liberty/com.ibm.websphere.wlp.doc/ae/rwlp_loose_applications.html[loose config^].
```

> The `liberty:run` argument specifies the Open Liberty `run` goal, which starts an Open Liberty instance in the foreground. As part of this phase, an Open Liberty runtime is downloaded and installed into the `target/liberty/wlp` directory, a Liberty instance is created and configured in the `target/liberty/wlp/usr/servers/defaultServer` directory, and the application is installed into that instance via https://www.ibm.com/support/knowledgecenter/en/SSEQTP_liberty/com.ibm.websphere.wlp.doc/ae/rwlp_loose_applications.html[loose config^].

Reviewer 2: ```suggestion
The `liberty:run` argument specifies the Open Liberty `run` goal, which starts an Open Liberty instance in the foreground. As part of this phase, an Open Liberty runtime is downloaded and installed into the `target/liberty/wlp` directory, a Liberty instance is created and configured in the `target/liberty/wlp/usr/servers/defaultServer` directory, and the application is installed into that instance using https://www.ibm.com/support/knowledgecenter/en/SSEQTP_liberty/com.ibm.websphere.wlp.doc/ae/rwlp_loose_applications.html[loose config^].
```

Author: updated

For more information about the Liberty Maven plug-in, see its https://github.com/WASdev/ci.maven[GitHub repository^].

> When the runtime instance begins starting up, various messages display in your command-line session. Wait for the following message, which indicates that the instance startup is complete:

Reviewer: ```suggestion
When the Liberty instance begins starting up, various messages display in your command-line session. Wait for the following message, which indicates that Liberty's startup is complete:
```

}
----

> When you need to stop the instance, press `CTRL+C` in the command-line session where you ran the instance, or run the `liberty:stop` goal from the `start` directory in another command-line session:

Reviewer: ```suggestion
When you need to stop the Liberty instance, press `CTRL+C` in the command-line session where you ran Liberty, or run the `liberty:stop` goal from the `start` directory in another command-line session:
```

// =================================================================================================

> // Starting and stopping the Open Liberty instance in the background

Reviewer 2: ```suggestion
// Starting and stopping Open Liberty in the background
```

// =================================================================================================

> == Starting and stopping the Open Liberty instance in the background

Reviewer 2: ```suggestion
== Starting and stopping Open Liberty in the background
```

> Although you can start and stop Liberty in the foreground by using the Maven `liberty:run` goal, you can also start and stop the instance in the background with the Maven `liberty:start` and `liberty:stop` goals:

Reviewer 2: ```suggestion
Although you can start and stop Liberty in the foreground by using the Maven `liberty:run` goal, you can also start and stop the Liberty instance in the background with the Maven `liberty:start` and `liberty:stop` goals:
```

> Although you can start and stop the instance in the foreground by using the Maven `liberty:run` goal, you can also start and stop the instance in the background with the Maven `liberty:start` and `liberty:stop` goals:

Reviewer: ```suggestion
Although you can start and stop Liberty in the foreground by using the Maven `liberty:run` goal, you can also start and stop the instance in the background with the Maven `liberty:start` and `liberty:stop` goals:
```

// =================================================================================================

> // Updating the runtime configuration without restarting the instance

Reviewer 2: ```suggestion
// Updating the runtime configuration without restarting Liberty
```

// =================================================================================================

> == Updating the runtime configuration without restarting the instance

Reviewer: ```suggestion
== Updating Liberty's configuration without restarting
```

> == Updating the runtime configuration without restarting the instance

Reviewer 2: ```suggestion
== Updating the runtime configuration without restarting Liberty
```

As before, you can see that the application is running by going to the http://localhost:9080/system/properties[http://localhost:9080/system/properties^] URL.

> Now try updating the runtime configuration while the instance is running in dev mode. The `system` microservice does not currently include health monitoring to report whether the instance and the microservice that it runs are healthy. You can add health reports with the MicroProfile Health feature, which adds a `/health` endpoint to your application. If you try to access this endpoint now at the http://localhost:9080/health/[http://localhost:9080/health/^] URL, you see a 404 error because the `/health` endpoint does not yet exist:

Reviewer: ```suggestion
Now try updating Liberty's `server.xml` configuration file while the instance is running in dev mode. The `system` microservice does not currently include health monitoring to report whether the instance and the microservice that it runs are healthy. You can add health reports with the MicroProfile Health feature, which adds a `/health` endpoint to your application. If you try to access this endpoint now at the http://localhost:9080/health/[http://localhost:9080/health/^] URL, you see a 404 error because the `/health` endpoint does not yet exist:
```

> Now try updating Liberty's `server.xml` configuration file while the instance is running in dev mode. The `system` microservice does not currently include health monitoring to report whether the instance and the microservice that it runs are healthy. You can add health reports with the MicroProfile Health feature, which adds a `/health` endpoint to your application. If you try to access this endpoint now at the http://localhost:9080/health/[http://localhost:9080/health/^] URL, you see a 404 error because the `/health` endpoint does not yet exist:

Reviewer 2: ```suggestion
Now try updating Liberty's `server.xml` configuration file while the instance is running in dev mode. The `system` microservice does not currently include health monitoring to report whether the Liberty instance and the microservice that it runs are healthy. You can add health reports with the MicroProfile Health feature, which adds a `/health` endpoint to your application. If you try to access this endpoint now at the http://localhost:9080/health/[http://localhost:9080/health/^] URL, you see a 404 error because the `/health` endpoint does not yet exist:
```

[source, role="no_copy"]
----
Error 404: java.io.FileNotFoundException: SRVE0190E: File not found: /health
----

> To add the MicroProfile Health feature to the instance, include the [hotspot=mpHealth]`mpHealth` feature in the [hotspot]`server.xml`.

Reviewer: ```suggestion
To add the MicroProfile Health feature to the Liberty instance, include the [hotspot=mpHealth]`mpHealth` feature in the [hotspot]`server.xml`.
```

[role="code_command hotspot", subs="quotes"]
----

> #Replace the runtime configuration file.#

Reviewer: ```suggestion
#Replace the `server.xml` configuration file.#
```

include::staging/server.xml[]
----

> After you make the file changes, Open Liberty automatically reloads its configuration. When enabled, the [hotspot=mpHealth]`mpHealth` feature automatically adds a `/health` endpoint to the application. You can see the instance being updated in the instance log displayed in your command-line session:

Reviewer: ```suggestion
After you make the file changes, Open Liberty automatically reloads its configuration. When enabled, the [hotspot=mpHealth]`mpHealth` feature automatically adds a `/health` endpoint to the application. You can see the instance being updated in the Liberty log displayed in your command-line session:
```

}
----

> Now you can verify whether your instance is up and running.

Reviewer: ```suggestion
Now you can verify whether your Liberty instance is up and running.
```

// =================================================================================================

> // Updating the source code without restarting the instance

Reviewer: ```suggestion
// Updating the source code without restarting Liberty
```

// =================================================================================================

> == Updating the source code without restarting the instance

Reviewer: ```suggestion
== Updating the source code without restarting Liberty
```

> The RESTful application that contains your `system` microservice runs in an instance from its `.class` file and other artifacts. Open Liberty automatically monitors these artifacts, and whenever they are updated, it updates the running instance without the need for the instance to be restarted.

Reviewer: ```suggestion
The RESTful application that contains your `system` microservice runs in a Liberty instance from its `.class` file and other artifacts. Open Liberty automatically monitors these artifacts, and whenever they are updated, it updates the running instance without the need for the instance to be restarted.
```

include::finish/pom.xml[tags=**]
----

> Try updating the source code while the instance is running in dev mode. At the moment, the `/health` endpoint reports whether the instance is running, but the endpoint doesn't provide any details on the microservices that are running inside of the instance.

Reviewer: ```suggestion
Try updating the source code while Liberty is running in dev mode. At the moment, the `/health` endpoint reports whether the Liberty instance is running, but the endpoint doesn't provide any details on the microservices that are running inside of the instance.
```

[INFO] [AUDIT] CWWKZ0003I: The application io.openliberty.guides.getting-started updated in 0.136 seconds.
----

> Access the `/health` endpoint again by going to the http://localhost:9080/health[http://localhost:9080/health^] URL. This time you see the overall status of your instance and the aggregated data of the liveness and readiness checks for the `system` microservice:

Reviewer: ```suggestion
Access the `/health` endpoint again by going to the http://localhost:9080/health[http://localhost:9080/health^] URL. This time you see the overall status of your Liberty instance and the aggregated data of the liveness and readiness checks for the `system` microservice:
```

You can also access the `/health/ready` endpoint by going to the http://localhost:9080/health/ready[http://localhost:9080/health/ready^] URL to view the data from the readiness health check. Similarly, access the `/health/live` endpoint by going to the http://localhost:9080/health/live[http://localhost:9080/health/live^] URL to view the data from the liveness health check.

> Making code changes and recompiling is fast and straightforward. Open Liberty dev mode automatically picks up changes in the `.class` files and artifacts, without needing to be restarted. Alternatively, you can run the `run` goal and manually repackage or recompile the application by using the `mvn package` command or the `mvn compile` command while the instance is running. Dev mode was added to further improve the developer experience by minimizing turnaround times.

Reviewer: ```suggestion
Making code changes and recompiling is fast and straightforward. Open Liberty dev mode automatically picks up changes in the `.class` files and artifacts, without needing to be restarted. Alternatively, you can run the `run` goal and manually repackage or recompile the application by using the `mvn package` command or the `mvn compile` command while Liberty is running. Dev mode was added to further improve the developer experience by minimizing turnaround times.
```

// =================================================================================================

> // Checking the Open Liberty instance logs

Reviewer: ```suggestion
// Checking the Open Liberty logs
```

// =================================================================================================

> == Checking the Open Liberty instance logs

Reviewer: ```suggestion
== Checking the Open Liberty logs
```

> While the instance is running in the foreground, it displays various console messages in the command-line session. These messages are also logged to the `target/liberty/wlp/usr/servers/defaultServer/logs/console.log` file. You can find the complete instance logs in the `target/liberty/wlp/usr/servers/defaultServer/logs` directory. The `console.log` and `messages.log` files are the primary log files that contain console output of the running application and the Liberty instance. More logs are created when runtime errors occur or whenever tracing is enabled. You can find the error logs in the `ffdc` directory and the tracing logs in the `trace.log` file.

Reviewer: ```suggestion
While Liberty is running in the foreground, it displays various console messages in the command-line session. These messages are also logged to the `target/liberty/wlp/usr/servers/defaultServer/logs/console.log` file. You can find the complete Liberty logs in the `target/liberty/wlp/usr/servers/defaultServer/logs` directory. The `console.log` and `messages.log` files are the primary log files that contain console output of the running application and the Liberty instance. More logs are created when runtime errors occur or whenever tracing is enabled. You can find the error logs in the `ffdc` directory and the tracing logs in the `trace.log` file.
```

[role="code_command hotspot" subs="quotes"]
----

> #Replace the runtime configuration file.#

Reviewer: ```suggestion
#Replace Liberty `server.xml` configuration file.#
```

Reviewer: "runtime configuration" may work.

Reviewer 2: Missing "the" so created another suggested commit.

Reviewer 2: I think your suggestion is fine otherwise.

> #Replace the runtime configuration file.#

Reviewer 2: ```suggestion
#Replace the Liberty `server.xml` configuration file.#
```

Updated to add "the".

[role="code_command hotspot file=0", subs="quotes"]
----

> #Replace the runtime configuration file.#

Reviewer: ```suggestion
#Replace the Liberty `server.xml` configuration file.#
```

Reviewer: "runtime configuration" may work.

== Running the application from a minimal runnable JAR

> So far, Open Liberty was running out of the `target/liberty/wlp` directory, which effectively contains an Open Liberty instance and the deployed application. The final product of the Maven build is a server package for use in a continuous integration pipeline and, ultimately, a production deployment.

Reviewer: ```suggestion
So far, Open Liberty was running out of the `target/liberty/wlp` directory, which effectively contains an Open Liberty installation and the deployed application. The final product of the Maven build is a server package for use in a continuous integration pipeline and, ultimately, a production deployment.
```

> Open Liberty supports a number of different server packages. The sample application currently generates a `usr` package that contains the runtime and application to be extracted onto an Open Liberty installation.

Reviewer: ```suggestion
Open Liberty supports a number of different server packages. The sample application currently generates a `usr` package that contains the Liberty runtime and application to be extracted onto an Open Liberty installation.
```

> Instead of creating a server package, you can generate a runnable JAR file that contains the application along with a runtime. This JAR file can then be run anywhere and deploy your application and runtime at the same time. To generate a runnable JAR file, override the  `include` property: 

Reviewer: ```suggestion
Instead of creating a server package, you can generate a runnable JAR file that contains the application along with a Liberty runtime. This JAR file can then be run anywhere and deploy your application and runtime at the same time. To generate a runnable JAR file, override the  `include` property: 
```

When the build completes, you can find the minimal runnable `guide-getting-started.jar` file in the `target` directory. This JAR file contains only the [hotspot=features file=1]`features` that you explicitly enabled in your [hotspot file=1]`server.xml` file. As a result, the generated JAR file is only about 50 MB.

> To run the JAR file, first stop the instance if it's running. Then, navigate to the `target` directory and run the `java -jar` command:

Reviewer: ```suggestion
To run the JAR file, first stop the Liberty instance if it's running. Then, navigate to the `target` directory and run the `java -jar` command:
```

[role='command']
```
java -jar guide-getting-started.jar
```

> When the instance starts, go to the http://localhost:9080/dev/system/properties[http://localhost:9080/dev/system/properties^] URL to access your application that is now running out of the minimal runnable JAR file.

Reviewer: ```suggestion
When Liberty starts, go to the http://localhost:9080/dev/system/properties[http://localhost:9080/dev/system/properties^] URL to access your application that is now running out of the minimal runnable JAR file.
```

> You can stop the instance by pressing `CTRL+C` in the command-line session that the instance runs in.

Reviewer: ```suggestion
You can stop the Liberty instance by pressing `CTRL+C` in the command-line session that the instance runs in.
```

== Great work! You're done!

> You've learned the basics of deploying and updating an application on an Open Liberty instance.

Reviewer: ```suggestion
You've learned the basics of deploying and updating an application on Open Liberty.
```
